### ISSUE TITLE
cluster expansiton fitting result is not good

### ISSUE THREAD

**Issue by shaliu19900911** [prisms-center/CASMcode#147]
Dear developer,

I tried first with 2 times of the primitive cell, and the fitting result seems good. But when I use larger supercell and more configurations, the fitting result seem worse. I include more cluster orbits and even tried enumerated parameters in the fit.json file, but the result is still bad. So I want to ask how this happen and how to solve this problem? 

2*primitive
![1](https://user-images.githubusercontent.com/46279371/54748866-30c65e00-4bd3-11e9-9074-dbb3b629db5d.png)

3*primitive
![1](https://user-images.githubusercontent.com/46279371/54748807-fbba0b80-4bd2-11e9-946a-fee4ab2ec819.png)

**Comment by jcthomas** [comment]
There are three possibilities that I typically consider when I encounter difficulty in fitting:

1) Tricky physics -- the configurational energies are determined by very long-range interactions that are difficult to account for with a cluster expansion. The cluster expansion can exactly reproduce a set of training data for a finite data set, if you have enough fitting parameters, so you might still get a good fit. 'Tricky physics' usually shows up when you have a fit for data 'up to supercell volume X' and then find that it is not accurate at supercell volume X+1. You may refit with the new data and then find that the accuracy is very bad at volume X+2. In other words, the fit never converges. **This doesn't seem to describe your situation.**

2) Underfitting -- If you get a bad fit you may be able to improve it by simply considering interactions at longer range (i.e., increasing the `"max_length"` in the `bspecs.json` file) or that involve more sites (i.e., adding more entries to `"orbit_branch_specs"` in the `bspecs.json` file for 3-body interactions, 4-body interactions, etc.). Alternatively, you can weight your fitting data so that energies near the convex hull are considered to be more important during fitting. This may help you get a fit that performs better at reproducing or predicting the convex hull, without adding many more interactions. The supported options for weighted fitting are described in the section about "problem_specs"/"weight" of the `casm-learn --settings-format` help screen.

3) Noisy data **(this seems the most likely explanation for the data you shared)** -- If your training dataset is sufficiently noisy, the underfitting strategies listed above won't help improve the fit. There are many potential causes of noise, but two that are common are **(a)** poor k-point convergence for DFT calculations and **(b)** large spontaneous relaxations. **In the case of (a)**, energies calculated in different supercells will be incompatible; you may be able to get a good fit for one or several small supercells, but as you add larger supercells to your training set, fits will perform increasingly poorly. The resolution is to increase the density of your KPOINT grid and recalculate energies for your training data. **In the case of (b)**, certain chemical orderings **may** cause the crystal to become mechanically unstable and undergo a large relaxation, potentially to a totally different parent crystal phase; if the parent crystal after relaxation is incompatible with `prim.json`, then the configuration should not be used for fitting the cluster expansion. 

You can query the magnitude of relaxations using `casm query -k "basis_deformation lattice_deformation sub(formation_energy,clex(formation_energy))" -c CALCULATED`, which will list the mean-square atomic displacement [Angstr^2/atom], the lattice deformation [Angstr^2/atom] and the fitting error for each calculated configuration. You can then explore configurations with large structural relaxations and/or fitting error to determine if they should be excluded from fitting. If there is another possible parent crystal structure for you chemistry, you can express it in a new prim file (.e., `alt_prim.json` and use casm to compare the mapping of each configuration onto `prim.json` vs onto `alt_prim.json`. This can be accomplished via `casm query -k "struc_score(prim.json) struc_score(alt_prim.json)"`. The column with the lower score indicates which parent crystal is a better match (an entry may be `unknown` if an incompatible mapping is attempted, like trying to map a 3-atom structure onto a 2-atom primitive structure).

**Comment by shaliu19900911** [comment]
Thanks very very much!! I have a deep look at DFT relaxed structure, and found the configurations that don't fit totally go to another structure after relaxation. So I think my case is in the case of 3(b).
